tensorflow/core/kernels/data/concatenate_dataset_op.cc

```diff
@@ -12,206 +12,209 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    TF_RETURN_IF_ERROR(
+        b->AddDataset(this, {input_graph, to_concatenate_graph}, output));
+    return Status::OK();
+  }
+
+ private:
+  class Iterator : public DatasetIterator<Dataset> {
+   public:
+    explicit Iterator(const Params& params)
+        : DatasetIterator<Dataset>(params), i_(0) {}
 
-    string DebugString() const override {
-      return "ConcatenateDatasetOp::Dataset";
+    Status Initialize(IteratorContext* ctx) override {
+      return dataset()->input_->MakeIterator(
+          ctx, strings::StrCat(prefix(), "[0]"), &input_impl_);
     }
 
-    int64 Cardinality() const override {
-      int64 n1 = input_->Cardinality();
-      int64 n2 = to_concatenate_->Cardinality();
-      if (n1 == kInfiniteCardinality || n2 == kInfiniteCardinality) {
-        return kInfiniteCardinality;
+    Status GetNextInternal(IteratorContext* ctx,
+                           std::vector<Tensor>* out_tensors,
+                           bool* end_of_sequence) override {
+      mutex_lock l(mu_);
+      if (!input_impl_) {
+        *end_of_sequence = true;
+        return Status::OK();
       }
-      if (n1 == kUnknownCardinality || n2 == kUnknownCardinality) {
-        return kUnknownCardinality;
+      while (i_ < 2) {
+        TF_RETURN_IF_ERROR(
+            input_impl_->GetNext(ctx, out_tensors, end_of_sequence));
+        if (!*end_of_sequence) {
+          return Status::OK();
+        }
+        if (++i_ < 2) {
+          TF_RETURN_IF_ERROR(dataset()->to_concatenate_->MakeIterator(
+              ctx, strings::StrCat(prefix(), "[1]"), &input_impl_));
+        }
       }
-      return n1 + n2;
+      *end_of_sequence = true;
+      input_impl_.reset();
+      return Status::OK();
     }
 
    protected:
-    Status AsGraphDefInternal(SerializationContext* ctx,
-                              DatasetGraphDefBuilder* b,
-                              Node** output) const override {
-      Node* input_graph = nullptr;
-      TF_RETURN_IF_ERROR(b->AddInputDataset(ctx, input_, &input_graph));
-      Node* to_concatenate_graph = nullptr;
-      TF_RETURN_IF_ERROR(
-          b->AddInputDataset(ctx, to_concatenate_, &to_concatenate_graph));
-      TF_RETURN_IF_ERROR(
-          b->AddDataset(this, {input_graph, to_concatenate_graph}, output));
-      return Status::OK();
+    std::shared_ptr<model::Node> CreateNode(
+        IteratorContext* ctx, model::Node::Args args) const override {
+      return model::MakeKnownRatioNode(std::move(args),
+                                       /*ratio=*/1);
     }
 
-   private:
-    class Iterator : public DatasetIterator<Dataset> {
-     public:
-      explicit Iterator(const Params& params)
-          : DatasetIterator<Dataset>(params), i_(0) {}
-
-      Status Initialize(IteratorContext* ctx) override {
-        return dataset()->input_->MakeIterator(
-            ctx, strings::StrCat(prefix(), "[0]"), &input_impl_);
+    Status SaveInternal(IteratorStateWriter* writer) override {
```

Comment:
let's also introduce name constants in the `Iterator ` class for the keys used in `SaveInternal` and `RestoreInteral` 
(e.g. "i" here and "input_impls_empty" in filter_dataset_op.cc) ... these do not need to be exposed in the header file

Reply:
Good suggestion. They are added.